x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/amazonbedrock/AmazonBedrockInferenceClientCache.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.amazonbedrock;
+
+import com.amazonaws.http.IdleConnectionReaper;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.inference.services.amazonbedrock.AmazonBedrockModel;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Collections.emptyMap;
+
+public final class AmazonBedrockInferenceClientCache implements AmazonBedrockClientCache {
+
+    private final BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator;
+    private volatile Map<Integer, AmazonBedrockBaseClient> clientsCache = emptyMap();
+
+    public AmazonBedrockInferenceClientCache(BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator) {
+        this.creator = creator;
+    }
+
+    public AmazonBedrockBaseClient getOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        var returnClient = internalGetOrCreateClient(model, timeout);
+        flushExpiredClients();
+        return returnClient;
+    }
+
+    private AmazonBedrockBaseClient internalGetOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        final Integer modelHash = AmazonBedrockInferenceClient.getModelKeysAndRegionHashcode(model, timeout);
+        {
+            final AmazonBedrockBaseClient client = clientsCache.get(modelHash);
+            if (client != null && client.tryToIncreaseReference()) {
+                return client;
+            }
+        }
+
+        synchronized (this) {
+            final AmazonBedrockBaseClient existing = clientsCache.get(modelHash);
+            if (existing != null && existing.tryIncRef()) {
+                return existing;
+            }
+
+            final AmazonBedrockBaseClient builtClient = creator.apply(model, timeout);
+
+            builtClient.mustIncRef();
+            clientsCache = Maps.copyMapWithAddedEntry(clientsCache, builtClient.hashCode(), builtClient);
+
+            return builtClient;
+        }
+    }
+
+    private synchronized void flushExpiredClients() {
+        var currentTimestampMs = System.currentTimeMillis();
+        for (final AmazonBedrockBaseClient client : clientsCache.values()) {
+            if (client.isExpired(currentTimestampMs)) {
+                client.decRef();
```

Comment:
If multiple threads call this one after the other without creating/accessing an expired client I think this will decrement the reference multiple times right? Even though it only expired "once".

Should we remove the client from the map if it is expired?

Thinking about this more, I wonder if it's possible to do this without reference counting 🤔 

I'm wondering what would happen if we removed an expired client from the map and call `internalClient.shutdown()`. I suppose if a client is about to be used after that, it'd fail. That'd probably be very unlikely though since when a client is retrieved a resets the 5 minute timer. So the client would have to be held for over 5 minutes. A way around that might be to return a wrapper of a client that has access to the cache and if a send call fails because the client is expired, the wrapper asks the cache for another client which would then create a new client.

I think that'd simplify having to remember to close a client after we're finished using it to send a request.


Another option would be to return a wrapper client that performs all access to the actual client in a `try-with-resource` so it immediately get's closed (reference decremented).

Reply:
> I think that'd simplify having to remember to close a client after we're finished using it to send a request.

I don't think that this is necessarily the behaviour we want - as we don't want to call the `shutdown()` method of the internal AWS client until its flushed from the cache (so it can be reused without recreating).  However, that being said - yeah - there should probably be something to at least decrease the reference...

> Thinking about this more, I wonder if it's possible to do this without reference counting

I'm starting to think along these lines as well... the reference counter is good to check and see if there's someone using the client - however, with the expiration being reset on each `get` of the client it shouldn't matter as long as there's an object reference in the cache...

> Should we remove the client from the map if it is expired?

That might be a good idea - I'll add this... and with this, we can remove the reference counting I think...

Reply:
I think this is the right way to go... I've updated my code.